src/main/kotlin/com/openlattice/data/storage/IndexingMetadataManager.kt

```diff
@@ -37,103 +86,121 @@ class IndexingMetadataManager(private val hds: HikariDataSource) {
-    val idColumn = if (linking) LINKING_ID.name else ID.name
-    val filterLinkingIds = if (linking) " AND ${LINKING_ID.name} IS NOT NULL " else ""
-    val entitiesClause = " AND ${ENTITY_SET_ID.name} = ? AND  $idColumn = ? $filterLinkingIds"
-    val withClause = buildWithClause(linking, entitiesClause)
-    // rather use id than linking_id in linking join
-    val joinClause = joinClause(IDS.name, FILTERED_ENTITY_KEY_IDS, entityKeyIdColumnsList)
-    val updateColumn = if (linking) LAST_LINK_INDEX.name else LAST_INDEX.name
+private val updateLastIndexSql = "UPDATE ${IDS.name} SET ${LAST_INDEX.name} = ? " +
+        "WHERE ${ENTITY_SET_ID.name} = ? AND ${ID.name} = ANY(?) AND ${PARTITION.name} = ?"
+
+/**
+ * Arguments of preparable sql in order:
+ * 1. last linking index
+ * 2. entity set id
+ * 3. linking ids (uuid array)
+ * 4. partitions (int array)
+ */
+private val updateLastLinkingIndexSql = "UPDATE ${IDS.name} SET ${LAST_LINK_INDEX.name} = ? " +
+        "WHERE ${ENTITY_SET_ID.name} = ? AND ${LINKING_ID.name} = ANY(?) " +
+        "AND ${LINKING_ID.name} IS NOT NULL AND ${PARTITION.name} = ANY(?)"
 
-    return "$withClause UPDATE ${IDS.name} SET $updateColumn = ? " +
-            "FROM $FILTERED_ENTITY_KEY_IDS WHERE ($joinClause)"
-}
 
 /**
  * Arguments of preparable sql in order:
- * 1. entity set ids (array)
+ * 1. entity set ids (uuid array)
+ * 2. partitions (int array)
  */
 fun markEntitySetsAsNeedsToBeIndexedSql(linking: Boolean): String {
-    val entitiesClause = " AND ${ENTITY_SET_ID.name} = ANY(?) "
-    val withClause = buildWithClause(linking, entitiesClause)
-    // rather use id than linking_id in linking join
-    val joinClause = joinClause(IDS.name, FILTERED_ENTITY_KEY_IDS, entityKeyIdColumnsList)
     val updateColumn = if (linking) LAST_LINK_INDEX.name else LAST_INDEX.name
 
-    return "$withClause UPDATE ${IDS.name} SET $updateColumn = '-infinity()' " +
-            "FROM $FILTERED_ENTITY_KEY_IDS WHERE ($joinClause) "
+    return "UPDATE ${IDS.name} SET $updateColumn = '-infinity()' " +
+            "WHERE ${VERSION.name} > 0 AND ${ENTITY_SET_ID.name} = ANY(?) AND ${PARTITION.name} = ANY(?)"
 }
 
 /**
  * Arguments of preparable sql in order:
- * 1. linking ids (array)
+ * 1. entity set id
+ * 2. partitions (int array)
+ * 3. linking ids (uuid array)
  */
-fun markLinkingIdsAsNeedToBeIndexedSql(): String {
-    val linkingEntitiesClause = " AND ${LINKING_ID.name} IS NOT NULL AND ${LINKING_ID.name} = ANY(?) "
-    val withClause = buildWithClause(true, linkingEntitiesClause)
-    // rather use id than linking_id in join
-    val joinClause = joinClause(IDS.name, FILTERED_ENTITY_KEY_IDS, entityKeyIdColumnsList)
-
-    return "$withClause UPDATE ${IDS.name} SET ${LAST_LINK_INDEX.name} = '-infinity()' " +
-            "FROM $FILTERED_ENTITY_KEY_IDS WHERE ($joinClause) "
-}
+private val markLinkingIdsAsNeedToBeIndexedSql = "UPDATE ${IDS.name} SET ${LAST_LINK_INDEX.name} = '-infinity()' " +
```

Comment:
Comment: version > 0 isn't necessary since marking a deleted entity as needing indexed won't have any effect and once if anything update the version it would qualify for needing indexed.